pkg/compactor/blocks_cleaner.go

```diff
@@ -685,24 +686,53 @@ var errStopIter = errors.New("stop iteration")
 // stalePartialBlockLastModifiedTime returns the most recent last modified time of a stale partial block, or the zero value of time.Time if the provided block wasn't a stale partial block
 func stalePartialBlockLastModifiedTime(ctx context.Context, blockID ulid.ULID, userBucket objstore.InstrumentedBucket, partialDeletionCutoffTime time.Time) (time.Time, error) {
 	var lastModified time.Time
-	err := userBucket.WithExpectedErrs(func(err error) bool {
-		return errors.Is(err, errStopIter) // sentinel error
-	}).Iter(ctx, blockID.String(), func(name string) error {
-		if strings.HasSuffix(name, objstore.DirDelim) {
+
+	var err error
+
+	// If bucket supports UpdatedAt IterOptionType, use IterWithAttributes
+	// to reduce the amount of object attributes calls.
+	if slices.Contains(userBucket.SupportedIterOptions(), objstore.UpdatedAt) {
+		err = userBucket.WithExpectedErrs(func(err error) bool {
+			return errors.Is(err, errStopIter) // sentinel error
+		}).IterWithAttributes(ctx, blockID.String(), func(attrs objstore.IterObjectAttributes) error {
+			if strings.HasSuffix(attrs.Name, objstore.DirDelim) {
+				return nil
+			}
+
+			modified, ok := attrs.LastModified()
+			if !ok {
+				// If for some reason we don't get the last modified time, we can't check it, so we just skip it.
+				return nil
+			}
```

Comment:
what are the possible reasons? how often would this happen?

Reply:
The answer is common prefixes, which would be captured by the `strings.HasSuffix(attrs.Name, objstore.DirDelim)` check above. This is a false case. Even without that it wouldn't be necessary because the `modified.After` checks would effectively skip it anyway since it would be a zero value of `time.Time`.